### DISCUSSION TITLE
Matplotlib figure of specific frame

### DISCUSSION THREAD

**Discussion by opherdonchin** [kineticstoolkit/kineticstoolkit#213] [2. Development: Ideas, new features, doc, etc.]
I want to make a cloud of multiple versions of a specific pose in order to communicate the variance in the different joints. This could build on existing code if the Player() object had an option that didn't automatically create an animation figure (off by default) and also had a `plot_frame()` method that could take an matplotlib axis object as one of its optional inputs. 

I don't mind trying to make this happen in a fork if you give me a couple of pointers of what and how to change things.

Thanks!
Opher

**Comment by felixchenier** [comment]
Hi @opherdonchin 

I think I understand what you want to do, but I want to be sure. Are you trying to superimpose different static poses on a same figure?

If this is what you need, then I see three different ways of doing it, without having to modify the Player class.

# 1. Probably your best bet: by using multiple TimeSeries

For this solution, I wrote a small sample code that could inspire you. It's based on the website sample code: https://kineticstoolkit.uqam.ca/doc/player.html

The trick is to build as many TimeSeries as there are postures to plot, and then to plot those all at once. The main difficulty with this is that we have to make every TimeSeries unique, with unique marker names, and do the same for the interconnections. Here, I plot every second of a sample TimeSeries:

```python
# %% Load sample data

import kineticstoolkit.lab as ktk

# Download and read markers from a sample C3D file
filename = ktk.doc.download("kinematics_tennis_serve.c3d")
markers = ktk.read_c3d(filename)["Points"]

# %% Define the interconnections

interconnections = dict()  # Will contain all segment definitions

interconnections["LLowerLimb"] = {
    "Color": [0, 0.5, 1],  # In RGB format (here, greenish blue)
    "Links": [  # List of lines that span lists of markers
        ["Derrick:LTOE", "Derrick:LHEE", "Derrick:LANK", "Derrick:LTOE"],
        ["Derrick:LANK", "Derrick:LKNE", "Derrick:LASI"],
        ["Derrick:LKNE", "Derrick:LPSI"],
    ],
}

interconnections["RLowerLimb"] = {
    "Color": [0, 0.5, 1],
    "Links": [
        ["Derrick:RTOE", "Derrick:RHEE", "Derrick:RANK", "Derrick:RTOE"],
        ["Derrick:RANK", "Derrick:RKNE", "Derrick:RASI"],
        ["Derrick:RKNE", "Derrick:RPSI"],
    ],
}

interconnections["LUpperLimb"] = {
    "Color": [0, 0.5, 1],
    "Links": [
        ["Derrick:LSHO", "Derrick:LELB", "Derrick:LWRA", "Derrick:LFIN"],
        ["Derrick:LELB", "Derrick:LWRB", "Derrick:LFIN"],
        ["Derrick:LWRA", "Derrick:LWRB"],
    ],
}

interconnections["RUpperLimb"] = {
    "Color": [1, 0.5, 0],
    "Links": [
        ["Derrick:RSHO", "Derrick:RELB", "Derrick:RWRA", "Derrick:RFIN"],
        ["Derrick:RELB", "Derrick:RWRB", "Derrick:RFIN"],
        ["Derrick:RWRA", "Derrick:RWRB"],
    ],
}

interconnections["Head"] = {
    "Color": [1, 0.5, 1],
    "Links": [
        ["Derrick:C7", "Derrick:LFHD", "Derrick:RFHD", "Derrick:C7"],
        ["Derrick:C7", "Derrick:LBHD", "Derrick:RBHD", "Derrick:C7"],
        ["Derrick:LBHD", "Derrick:LFHD"],
        ["Derrick:RBHD", "Derrick:RFHD"],
    ],
}

interconnections["TrunkPelvis"] = {
    "Color": [0.5, 1, 0.5],
    "Links": [
        ["Derrick:LASI", "Derrick:STRN", "Derrick:RASI"],
        ["Derrick:STRN", "Derrick:CLAV"],
        ["Derrick:LPSI", "Derrick:T10", "Derrick:RPSI"],
        ["Derrick:T10", "Derrick:C7"],
        ["Derrick:LASI", "Derrick:LSHO", "Derrick:LPSI"],
        ["Derrick:RASI", "Derrick:RSHO", "Derrick:RPSI"],
        [
            "Derrick:LPSI",
            "Derrick:LASI",
            "Derrick:RASI",
            "Derrick:RPSI",
            "Derrick:LPSI",
        ],
        [
            "Derrick:LSHO",
            "Derrick:CLAV",
            "Derrick:RSHO",
            "Derrick:C7",
            "Derrick:LSHO",
        ],
    ],
}


# =============================
# THIS IS WHERE THE WORK BEGINS
# =============================


# %% See every frame at 0, 1, 2, 3, 4, 5, 6, 7 seconds

timeseries_to_plot = []  # List of TimeSeries
all_interconnections = {}

for t in [0, 1, 2, 3, 4, 5, 6, 7]:
    # -------------------------------------------
    # Build a TimeSeries of one frame at time = t
    # -------------------------------------------

    # Cut everything before t
    temp_ts = markers.get_ts_after_time(t, inclusive=True)
    # Keep only one index
    temp_ts = temp_ts.get_ts_before_index(1)
    # Set this time to 0 so that every TimeSeries ends at t=0
    temp_ts.shift(-t, in_place=True)

    # -------------------------------------------
    # Modify the marker names so that everything is unique
    # -------------------------------------------

    # Do this for the TimeSeries
    marker_names = list(temp_ts.data.keys())
    for marker_name in marker_names:
        temp_ts.rename_data(marker_name, f"{marker_name}_{t}", in_place=True)

    timeseries_to_plot.append(temp_ts)

    # Do this for the interconnections

    # Go through every body segment
    for body_name in interconnections:
        all_interconnections[f"{body_name}_{t}"] = dict()
        all_interconnections[f"{body_name}_{t}"]["Color"] = interconnections[
            body_name
        ]["Color"]
        all_interconnections[f"{body_name}_{t}"]["Links"] = []

        # Go through every link of this segment
        for i_link, link in enumerate(interconnections[body_name]):
            all_interconnections[f"{body_name}_{t}"]["Links"].append(
                [
                    f"{s}_{t}"
                    for s in interconnections[body_name]["Links"][i_link]
                ]
            )

# %% Plot all these TimeSeries in a single Player window

ktk.Player(*timeseries_to_plot, interconnections=all_interconnections, up="z")
```

The result:

![Screen Shot 2023-12-05 at 09 01 17](https://github.com/kineticstoolkit/kineticstoolkit/assets/34967663/160f59a5-9888-424c-b3dc-0cf888364a28)

Obviously you will have lots of editing to make this work for your data, but it gives you the idea.

# 2. Hacking the Player's figure data

This is not what I'd recommend for long term use because it means that you play in the undocumented implementation and it may break at any time. But in case option 1 does not work well for you, you can access the inner objects created by a player using its private `objects` argument.

```python
player = ktk.Player(*timeseries_to_plot, interconnections=all_interconnections, up="z")

player.objects
```

```
{
             'PlotMarkers': <dict with 14 entries>
        'PlotRigidBodiesX': <matplotlib.lines.Line2D object at 0x1799e36a0>
        'PlotRigidBodiesY': <matplotlib.lines.Line2D object at 0x1799e3940>
        'PlotRigidBodiesZ': <matplotlib.lines.Line2D object at 0x1799e3be0>
         'PlotGroundPlane': <matplotlib.lines.Line2D object at 0x1799e3400>
    'PlotInterconnections': <dict with 48 entries>
                  'Figure': <Figure size 900x719 with 1 Axes>
                    'Axes': <Axes: title={'center': 'Player'}>
                    'Help': None
}
```

The figure and its contents is directly accessible as `player.objects["Figure"]`. The markers plots and interconnections plots are also accessible as Line2D objects in `player.objects["PlotMarkers"]` and `player.objects["PlotInterconnections"]`. You could theoretically copy this information to another figure that you build in parallel.

# 3. Not using the Player

The Player is built as a way to visualize animations, and lots of work has been done in getting an acceptable performance with Matplotlib. As such, it is not easily extendable. But if animations is not what you want, then you could also use the Matplotlib's 3d plot functions directly. It would give you total control on the look of your output.

In any case, I'm curious which solution you will use, and also if I understand your problem well. I'd be happy if you made me a follow up.

Best,

Félix

**Comment by opherdonchin** [comment]
Hi @felixchenier 

Thanks for the very quick reply! From the code you wrote, it seems you did understand exactly what I was thinking about. Also, from what you wrote, it seems like `Player` doesn't really add much simplicity (in its current form). That was sort of the conclusion that I reached. 

My code is below. I think everything here is pretty self-explanatory except that:

- We are plotting the pose of a person making a billiards shot at the time the cue ball is hit. We want to overlay multiple poses to demonstrate what about the pose varies from shot to shot and what stays the same.
- `marker_data['shots']` is a list of dictionaries that each contain a time series with key `ts` that has the marker data. I want to plot the pose at time 0 for several of these.
- The list of markers for the actual person is in `consts.body_markers` (but I need to get rid of a few)
- The list of makers for the pool cue the person is holding is in `consts.cue_markers`

I just brute-forced this. It seems to me it would be elegant to add a function to `TimeSeries` or `Player` that would take a time series, an event or time or index, and interconnections, and plot the pose into the current figure or a new one.

Anyway here is my code, followed by the figure it generated.

Will be glad for any further thoughts.
Opher

```
num_to_plot = 15  # Adjust as needed
total_shots = len(marker_data['shots'])
selected_shots = np.linspace(0, total_shots - 1, num_to_plot, dtype=int)


fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')

for shot in selected_shots:
    ts = marker_data['shots'][shot]['ts']

    body_markers = [m for m in consts.body_markers if not re.compile(r'(R|L)_(ELB|W)_(L|M)').match(m)]
    plot_ts = ts.get_subset(body_markers)


    success_test_index = ts.get_index_before_time(0)

    for joint_name, joint_data in plot_ts.data.items():
        pt = joint_data[success_test_index]
        ax.scatter3D(xs=pt[0], ys=pt[1], zs=pt[2], c='black')

    plot_ts = ts.get_subset(consts.body_markers + consts.cue_markers)

    for part,part_dict in interconnections.items():
        color = part_dict['Color']
        links = part_dict['Links']
        for link in links:
            pts = [plot_ts.data[joint][success_test_index, 0:3] for joint in link]
            pts = np.array(pts)
            ax.plot3D(pts[:,0], pts[:,1], pts[:,2], c=color)

ax.azim = 30
ax.elev = 0
ax.set_box_aspect(N
one, zoom=1.5)
```

![Figure_7](https://github.com/kineticstoolkit/kineticstoolkit/assets/26271102/2f8054b1-310b-41a9-b1e0-dbafdbbed7bd)

**Reply by felixchenier** [reply]
Thanks @opherdonchin for your detailed answer. Yes, you're right, this feature would be nice to have.

I think I would keep the idea of the Player, with its familiar black background and mouse gestures, but I'd add a functionality a bit like the `plt.hold(True)` or `plt.hold(False)` of Matplotlib, where we can add stuff on top of the current stuff. Maybe something like:

```
player = ktk.Player(markers1, interconnections=interconnections)
player.add(markers2, interconnections=interconnections)
player.add(markers3, interconnections=interconnections)
```

and so on. This function would take care of marker renaming (adding TimeSeries may add `_1`, `_2`, `_n` suffixes to the markers so that each marker stays unique. By keeping the Player instead of converting to a Matplotlib window, we also keep the animation system. This leads to other potentially useful things, such as comparing different athletes side-by-side or superposed, or techniques, or anything. I'll take some time to think about it. It wouldn't be that complex to program, I just need to be sure to get the interface and expected behaviour right.

A big thank you for the idea and I'm glad that you reached your objective.

**Reply by opherdonchin** [reply]
Yes! I think everything you are saying makes very good sense! I really like the idea of making this possible in the player to allow overlapped animations. 

Here are some more thoughts on this:

- These should be `ts` and not `markers`, right?
- If there were `time=time`, `event=event`, and/or `index=index` parameters for the `ktk.Player`  and `add` functions, you could select the snapshot you want. 
- Could one add an `ax=` argument to the constructor which would toggle between the curent behavior with background, grid and animations and a simpler plot which only adds the markers/interconnections to the axes? I don't know enough about the details.
-  Defaulting to `_1`, `_2`, `_n` makes sense, but you could also allow the `ktk.Player` and `add` functions to take an optional `name=` argument to allow specific naming. 

If there is anything in this whole idea which would be possible to work on as a first step, I'd be sort of excited to try to help out. 

I'm guessing, but might the first step be refactoring the plotting of the markers/interconnections into a separate `_draw_markers` function? 

Opher

**Reply by felixchenier** [reply]
@opherdonchin See answer below for the rest of this discussion 😃

For your specific questions:

> These should be ts and not markers, right?

Yes.

> If there were time=time, event=event, and/or index=index parameters for the ktk.Player and add functions, you could select the snapshot you want.

Better yet, in this new issue: https://github.com/kineticstoolkit/kineticstoolkit/issues/215 we have read/write properties to navigate in the Player without needing it to be interactive.

Using this API, we could show a same TimeSeries at different events using something like:

```python
for i in ts.count_events("event"):
    temp_ts = ts.get_ts_after_event("event", i)
    temp_ts.shift(temp_ts.time[0], in_place=True)  # This event is now at time 0
    player.add(temp_ts, name=f"Event{i}")
```
Done!

> Could one add an ax= argument to the constructor which would toggle between the curent behavior with background, grid and animations and a simpler plot which only adds the markers/interconnections to the axes? I don't know enough about the details.

I made a detailed explanation in the issue on why I don't think this is a good idea, at least for short to middle term.

> Defaulting to _1, _2, _n makes sense, but you could also allow the ktk.Player and add functions to take an optional name= argument to allow specific naming.

Yes! Excellent idea. I used this idea for the first draft of API. I really like this idea to "group" TimeSeries by name. It also allows removing TimeSeries from the Player, using the name.

> If there is anything in this whole idea which would be possible to work on as a first step, I'd be sort of excited to try to help out.

Let's start with commenting the draft of API. We want it to address your needs without restricting future improvement by using a too narrow scope.

> I'm guessing, but might the first step be refactoring the plotting of the markers/interconnections into a separate _draw_markers function?

Just wait after I clean up this code. Many things will have to change anyway, and I'll make sure to refactor the code in a way that allows these future changes.

**Comment by felixchenier** [comment]
Hi @opherdonchin 

I really like where this discussion is going, and this may be the starting point for something I envisioned long ago, but that wasn't on the priority list yet:

# A scripting API for the Player

Instead of having loads of options on the constructor, I think an API with well documented rw/ro properties would be still more versatile and helpful. This also means more work, but in the long term it will be great.

The first step is on my side: I need to cleanup the Player's code before we start fiddling with it. It was one of the first modules I programmed in KTK and it can be simplified a lot.

The second step (that can be done in parallel) is to establish the API and feature set. Obviously you can help me there, because you already do.

The third step will be to program it. At this point, I'll be happy if you want to fork and participate, and I could work on it also.

I'm all in for this scripting API, therefore I create an issue from this discussion, where we can move on to do it.
